Decay heat model

smpark7 opened the pull request:
Closes #47 

New:
- HeatPrecursorSource
- HeatPrecursorDecay
- CoupledOutflowBC
- PostprocessorCoupledInflowBC
- SideCoupledIntegralVariablePostprocessor
- Tests for HeatPrecursorSource and HeatPrecursorDecay

Modified:
- TransientFissionHeatSource
- PrecursorAction

This PR introduces new kernels for decay heat precursor source and decay. The theoretical formulation for the decay heat model is:
![image](https://user-images.githubusercontent.com/30237775/93461408-4aa34700-f917-11ea-9c94-ada8f94eaf24.png)

Accordingly, I've edited the ```TransientFissionHeatSource``` kernel to work with the decay heat model. The heat source has an additional contribution from decay heat as shown here:
![image](https://user-images.githubusercontent.com/30237775/93461794-df0da980-f917-11ea-91e7-ec83dc2f98dd.png)

Users can activate/deactivate decay heat modeling in ```TransientFissionHeatSource``` using the ```account_decay_heat``` boolean flag.

I've also written a test for the decay heat kernels that is similar to the existing test for delayed neutron precursors. It solves for the steady-state decay heat source distribution in the channel-type reactor geometry under a fixed uniform two-group neutron flux.

```CoupledOutflowBC```, ```PostprocessorCoupledInflowBC```, and ```SideCoupledIntegralVariablePostprocessor``` are analogous to ```OutflowBC```, ```PostprocessorInflowBC```, and ```SideAverageValue``` for setting boundary conditions for delayed neutron precursors and decay heat precursors for a MultiApp solve involving a main core geometry and a 1D outer loop geometry for looping the precursors. The difference is that the new kernels support the use of MOOSE Navier-Stokes velocities instead of constant uniform velocities.

As such, ```PrecursorAction``` has been edited to include these kernels and users can now use MOOSE Navier-Stokes velocities with the precursor looping MultiApp.

gridley commented:
Man, this is awesome. I think it will be a pretty important effect to consider for transients, and I am not sure of any stuff in the literature that quantifies the impact of this drifting effect. Everyone talks about the precursors moving, but 7% of the heat is pretty serious... Especially for prompt supercritical stuff, this will be really interesting.

Anyways, I glanced at the code changes and just wanted to point out something that's not immediately obvious if you're working off of slightly older MOOSE code. They changed the style of `validParams` at some point so it's a good bit cleaner now. Of course, both syntaxes work, it's just that the newer one is cleaner. Rather than defining a template specialization of `validParams` on the class you're making, you can now just define a static method in your class that returns the parameters for that class. I recommend checking out how they set up the `HeatConduction` kernel in the module of the same name.

robfairh commented on src/postprocessors/SideCoupledIntegralVariablePostprocessor.C:
```diff
@@ -0,0 +1,39 @@
+Real
+SideCoupledIntegralVariablePostprocessor::computeQpIntegral()
+{
+  return _u[_qp] * _weight[_qp] / _divisor;
```
Is there a way to make sure that _divisor is never 0?

smpark7 replied:
Yea I'll include a MooseError call there to check whether it's 0

smpark7 replied:
I've changed reworked this postprocessor to make it easier to understand and use. It now works in conjunction with DivisionPostprocessor to provide flow-averaged quantities or other uses.

robfairh reviewed (commented):
I've checked the following files so far:
* include/bcs/CoupledOutflowBC.h
* include/bcs/PostprocessorCoupledInflowBC.h
* include/postprocessors/SideCoupledIntegralVariablePostprocessor.h
* src/bcs/CoupledOutflowBC.C
* src/bcs/PostprocessorCoupledInflowBC.C
* src/postprocessors/SideCoupledIntegralVariablePostprocessor.C

I feel that all those files should have been in a different PR.
In the future consider making smaller PRs.
I will review the decay heat decay related kernels soon.

robfairh commented on src/kernels/HeatPrecursorSource.C:
```diff
@@ -0,0 +1,95 @@
+  params.addRequiredCoupledVar("group_fluxes", "All the variables that hold the group fluxes. "
+                                               "These MUST be listed by decreasing "
+                                               "energy/increasing group number.");
+  params.addParam<Real>("nt_scale", 1, "Scaling of the neutron fluxes to aid convergence.");
```
yes

robfairh commented on src/bcs/CoupledOutflowBC.C:
```diff
@@ -0,0 +1,52 @@
+validParams<CoupledOutflowBC>()
+{
+  InputParameters params = validParams<IntegratedBC>();
+  params.addRequiredCoupledVar("uvel", "x direction velocity");
```
This is not a common practice in moose. It doesn't define a default value because it is a required variable for the simulation to run. If the variable has not been defined, the simulation won't run. So, defining a default value is redundant.

smpark7 replied:
I didn't provide a default value because it would inform the user that he/she/they did not provide a coupled velocity variable. We have the existing OutflowBC.C for constant velocity values.

smpark7 commented on src/bcs/PostprocessorCoupledInflowBC.C:
```diff
@@ -0,0 +1,48 @@
+validParams<PostprocessorCoupledInflowBC>()
+{
+  InputParameters params = validParams<IntegratedBC>();
+  params.addRequiredCoupledVar("uvel", "x direction velocity");
```
Same explanation as the CoupledOutflowBC case.

smpark7 commented on src/kernels/DecayHeatSource.C:
```diff
@@ -0,0 +1,66 @@
+{
+  InputParameters params = validParams<Kernel>();
+  params += validParams<ScalarTransportBase>();
+  params.addRequiredParam<unsigned int>("num_decay_heat_groups", "The number of decay heat groups.");
```
I can see users using default values as a rough estimate for decay heat but I'm not sure how appropriate it is. The 2014 ANSI standard uses 23 groups for decay heat which would really impact performance. The 3-group data I used for the MSFR is specifically with the U233-Th-Li-F fuel composition for t = 0s to 300s after long operation.

katyhuff commented:
Please merge most recent devel branch (and accordingly trigger CIVET tests).

smpark7 commented on src/actions/PrecursorAction.C:
```diff
@@ -324,23 +327,50 @@ PrecursorAction::bcAct(const std::string & var_name)
-    params.set<Real>("ww") = getParam<Real>("w_def");
-    params.set<PostprocessorName>("postprocessor") =
-        "Inlet_SideAverageValue_" + var_name + "_" + _object_suffix;
+    if (getParam<bool>("constant_velocity_values"))
```
I've created a separate issue #139 to overhaul the code structure in PrecursorAction.C for better readability

smpark7 commented on src/actions/PrecursorAction.C:
```diff
@@ -373,15 +403,30 @@ PrecursorAction::postAct(const std::string & var_name)
-    _problem->addPostprocessor("SideAverageValue", postproc_name, params);
+    if (isParamValid("uvel"))
+    {
+      InputParameters params = _factory.getValidParams("SideCoupledIntegralVariablePostprocessor");
```
I've created a separate issue #139 to overhaul the code structure in PrecursorAction.C for better readability

smpark7 commented on src/actions/PrecursorAction.C:
```diff
@@ -80,6 +80,9 @@ validParams<PrecursorAction>()
   params.addParam<bool>("is_loopapp", "if circulating precursors, whether this is loop app");
   params.addParam<bool>("eigen", false, "whether neutronics is in eigenvalue calculation mode");
+  params.addParam<NonlinearVariableName>("weight",
+                                         "The name of the variable that the SideCoupledIntegralVariablePostprocessor is integrated with");
```
I've reworked the postprocessor and PrecursorAction to be less confusing and more intuitive. The user is now required to provide the flow velocity variable normal to the outlet boundary.

smpark7 commented:
Thank you for your comments @katyhuff @robfairh 

This PR is ready for a new review. To recap, this PR introduces a decay heat model in Moltres, and supporting BCs and postprocessors for coupling Moltres' capabilities with the MOOSE Navier-Stokes module. Here's a [pdf](https://github.com/arfc/moltres/files/5493077/report.pdf) of the derivation for the decay heat model and the various assumptions used. Users provide ```num_decay_heat_groups```, ```decay_heat_fractions```, and ```decay_heat_constants```, which correspond to the number of decay heat groups, the fractions of decay heat to full power associated to each group, and the decay constants associated to each group, respectively.

The ```HeatPrecursorSource``` and ```HeatPrecursorDecay``` kernels correspond to the source and decay terms for the decay heat precursors.

The pre-existing```TransientFissionHeatSource``` kernel now includes an ```account_decay_heat``` flag which subtracts the total fraction of the fission power allocated to decay heat from the prompt fission power.

```CoupledOutflowBC``` and ```PostprocessorCoupledInflowBC``` calculate correct DG residuals and jacobians for the inflow and outflow of precursors.

```SideWeightedIntegralPostprocessor``` calculates the flow-averaged precursor concentration outflow to be fed into ```PostprocessorCoupledInflowBC``` when looping the precursors between the main app and the loop app.

I've also incorporated the BCs and postprocessor above into ```PrecursorAction``` to fully support using Navier-Stokes velocities for looping delayed neutron precursors. 

Lastly, I've added tests for the decay heat capability, the coupled BCs, the ```SideWeightedIntegralPostprocessor```, and looping delayed neutron precursors with Navier-Stokes velocities.

katyhuff reviewed (approved):
Good job!
